### DISCUSSION TITLE
Differentiation

### DISCUSSION THREAD

**Discussion by saulshanabrook** [Quansight-Labs/metadsl#129] [Ideas]
We should be able to compute derivatives of array expression wrt another array expression. 

Prior work:

* http://www.matrixcalculus.org/matrixcalculus.pdf
* http://conal.net/papers/essence-of-ad/
* http://jmlr.org/papers/volume18/17-468/17-468.pdf

**Comment by dhirschfeld** [comment]
Perhaps there are any takeaways from [`autograd`](https://github.com/HIPS/autograd) or [`tangent`](https://github.com/google/tangent)?

Also from autograd:
 - https://github.com/HIPS/autograd/blob/master/docs/tutorial.md
 - https://dougalmaclaurin.com/phd-thesis.pdf

I really like autograd, probably because it adheres closely to the numpy api, however performance can be an issue.

It would be amazing if uarray could compute the derivatives in a way which could be compiled by numba and scheduled on a distributed dask cluster!

***xref:*** https://github.com/HIPS/autograd/pull/397

**Comment by saulshanabrook** [comment]
SymPy is implementing a new matrix differentiation https://github.com/sympy/sympy/pull/15615 based on http://www.matrixcalculus.org/

**Comment by asmeurer** [comment]
The [matrix cookbook](https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf) is also a good reference to be aware of here.

**Comment by Upabjojr** [comment]
> SymPy is implementing a new matrix differentiation [sympy/sympy#15615](https://github.com/sympy/sympy/pull/15615) based on http://www.matrixcalculus.org/

Not really based on the algorithm of MatrixCalculus.org. I did not implement neither a DAG nor Ricci calculus for matrix derivatives.
